### PR TITLE
Fix for ISSUE #153069

### LINKED ISSUE #153069: `torch.ldexp` goes out of range when `2**other` is out of range
### 🐛 Describe the bug

```python
import torch
torch.ldexp(torch.tensor([2], dtype=torch.float16), torch.tensor([-25], dtype=torch.int32))
```
Gives
```python
tensor([0.], dtype=torch.float16)
```
Even though `2 * 2**-25 = 2**-24` is non-zero and is within representable range of `torch.float16`, and
```python
torch.ldexp(torch.tensor([1], dtype=torch.float16), torch.tensor([-24], dtype=torch.int32))
```
correctly outputs
```python
tensor([5.9605e-08], dtype=torch.float16)
```
I'm not sure if this is WAI, or if there's a way to still make the first example output the correct answer without hurting performance? Thanks!

### Versions

2.6.0

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim

### PR DESCRIPTION
**Pull request by aritra3520**
Fixes #153069

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

### CODE REVIEW

**Review by ngimel — COMMENTED** [review]
This provides only cpu support (and ldexp on cuda will now throw an error). You need to create a stub that would correctly dispatch to cpu or cuda implementation.

**Comment by aritra3520** [comment]
> This provides only cpu support (and ldexp on cuda will now throw an error). You need to create a stub that would correctly dispatch to cpu or cuda implementation.

Ok yes i will create a stub for proper device management. Thank you

**Comment by aritra3520** [comment]
Hi @ngimel do you expect cuda implementation too or just declaring a stub will do?

**Comment by ngimel** [comment]
Of course you need cuda implementation, you can't regress current coverage.

File: aten/src/ATen/native/cuda/BinaryMiscOpsKernels.cu [hunk 1]
```diff
@@ -69,12 +69,35 @@ void xlog1py_kernel_cuda(TensorIteratorBase& iter) {
   });
 }
 
+// Note:
+// For low-precision dtypes like float16, results may overflow or underflow.
+// - For very small exponents (e.g., exp <= -25), the result will likely underflow to 0.0.
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/BinaryMiscOpsKernels.cu:74]: this is not true because if x is large it shouldn't underflow?

File: test/test_torch.py [hunk 2]
```diff
@@ -10977,6 +10977,72 @@ def add_neg_dim_tests():
         assert not hasattr(TestTorch, test_name), "Duplicated test name: " + test_name
         setattr(TestTorch, test_name, make_neg_dim_test(name, tensor_arg, arg_constr, types, extra_dim))
 
+def test_ldexp_pow2_float16_cases():
```
Comment (ngimel) [inline test/test_torch.py:10980]: add the test to test_binary_ufuncs.py instead of here, parametrize it on device type instead of hardcoding device

File: test/test_torch.py [hunk 3]
```diff
@@ -10977,6 +10977,72 @@ def add_neg_dim_tests():
...
+            print(f"  In-place: {a_clone.item()} (expected: {expected.item()})")
+            assert torch.allclose(a_clone, expected, atol=1e-8) or \
+                   (torch.isinf(expected) and torch.isinf(a_clone)), \
+                   f"(in-place) Expected {expected.item()}, got {a_clone.item()}"
+
+            # Out variant
+            out_tensor = torch.empty_like(a)
+            torch.ldexp(a, b, out=out_tensor)
+            print(f"  Out: {out_tensor.item()} (expected: {expected.item()})")
+            assert torch.allclose(out_tensor, expected, atol=1e-8) or \
```
Comment (ngimel) [inline test/test_torch.py:11042]: self.assertEqual (test should inherit from TestCase and be a method in the test class)

File: test/test_torch.py (hunk 2 above)
Comment (aritra3520) [reply test/test_torch.py:10980]: Yes I will fix this issue and update the PR

File: test/test_torch.py (hunk 3 above)
Comment (aritra3520) [reply test/test_torch.py:11042]: Yes I will fix this issue and update the PR

Comment (aritra3520) [reply test/test_torch.py]: Hi I did the fix can you review. Thank you

File: test/test_torch.py (hunk 2 above)
Comment (aritra3520) [reply test/test_torch.py]: Hi I did the fix can you review. Thank you

File: aten/src/ATen/native/BinaryOps.cpp [hunk 4]
```diff
@@ -1542,27 +1551,29 @@ TORCH_IMPL_FUNC(heaviside_out) (
...
+    .add_input(other)
+    .build();
+
+  ldexp_stub(result.device().type(), iter); 
+  return result;
+}
 
 Tensor ldexp(const Tensor& self, const Tensor& other) {
-  return at::mul(self, _pow2(self, other));
+    auto result_dtype = at::result_type(self, other);
```
Comment (ngimel) [inline aten/src/ATen/native/BinaryOps.cpp:1570]: This still won't work:
```
In [65]: x=torch.ones(4, dtype=torch.int)

In [66]: y=torch.ones(4, dtype=torch.int)

In [67]: torch.ldexp(x, y).dtype
Out[67]: torch.float32

In [68]: torch.result_type(x, y)
Out[68]: torch.int32
```

File: aten/src/ATen/native/BinaryOps.cpp [hunk 5]
```diff
@@ -1542,27 +1551,29 @@ TORCH_IMPL_FUNC(heaviside_out) (
...
-      return at::pow(2.0, other);
-  }
-  // For double and reduced floating types do regular type promotion
-  return at::full({}, 2.0, self.options()).pow(other);
-}
-
 Tensor& ldexp_out(const Tensor& self, const Tensor& other, Tensor& result) {
-  return at::mul_out(result, self, _pow2(self, other));
-}
+  TORCH_CHECK(other.scalar_type() == at::kInt || other.scalar_type() == at::kLong,
```
Comment (ngimel) [inline aten/src/ATen/native/BinaryOps.cpp:1555]: I'm not sure this will be correct if you send output of unexpected dtype (integer or complex or double when float is expected)

File: aten/src/ATen/native/BinaryOps.cpp (hunk 4 above)
Comment (aritra3520) [reply aten/src/ATen/native/BinaryOps.cpp:1570]: So you are suggesting that the output should always be float.

File: aten/src/ATen/native/BinaryOps.cpp (hunk 5 above)
Comment (aritra3520) [reply aten/src/ATen/native/BinaryOps.cpp]: Ya I made the updates so that it only gives float outputs
I included a Torch_check for that

File: aten/src/ATen/native/BinaryOps.cpp (hunk 4 above)
Comment (ngimel) [reply aten/src/ATen/native/BinaryOps.cpp]: Check what are the current output dtypes. I think it's double when one of the inputs is double, or complex if one of the inputs is complex, following type promotion rules (e.g. if one of the inputs is 0-dim tensor then its dtype has a lower priority)

Comment (aritra3520) [reply aten/src/ATen/native/BinaryOps.cpp]: I saw the type promotion and updated the fix. can you review

**Review by ngimel — COMMENTED** [review]
There are a lot of failing tests

**Comment by aritra3520** [comment]
> Ok a lot of test errors are due to sample inputs for ldexp operation not being prepared correctly, I guess you'd need to fix that by adding a custom `sample_inputs_func` here https://github.com/pytorch/pytorch/blob/edc2d539d1e8040065b6a0f83638218e16d44283/torch/testing/_internal/common_methods_invocations.py#L13986

Will surely look into this thank you
